File: packages/camera/camera_avfoundation/example/ios/RunnerTests/Mocks/MockAssetWriter.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "MockAssetWriter.h"

@implementation MockAssetWriter

- (BOOL)startWriting {
  if (self.startWritingStub) {
    self.startWritingStub();
  }
  return YES;
}

- (AVAssetWriterStatus)status {
  if (_statusStub) {
    return _statusStub();
  }
  return AVAssetWriterStatusUnknown;
}

- (void)finishWritingWithCompletionHandler:(void (^)(void))handler {
  if (self.finishWritingStub) {
    self.finishWritingStub(handler);
  }
}

- (void)startSessionAtSourceTime:(CMTime)startTime {
}

- (void)addInput:(AVAssetWriterInput *)input {
}

@end

@implementation MockAssetWriterInput

- (BOOL)appendSampleBuffer:(CMSampleBufferRef)sampleBuffer {
  if (self.appendSampleBufferStub) {
    return self.appendSampleBufferStub(sampleBuffer);
  }
  return YES;
}

@end

@implementation MockAssetWriterInputPixelBufferAdaptor

- (BOOL)appendPixelBuffer:(CVPixelBufferRef)pixelBuffer
     withPresentationTime:(CMTime)presentationTime {
  if (self.appendPixelBufferStub) {
    return self.appendPixelBufferStub(pixelBuffer, presentationTime);
  }
  return YES;
}

@end

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/Mocks/MockAssetWriter.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@testable import camera_avfoundation

// Import Objective-C part of the implementation when SwiftPM is used.
#if canImport(camera_avfoundation_objc)
  import camera_avfoundation_objc
#endif

/// Mock implementation of `AssetWriter` protocol which allows injecting a custom
/// implementation.
final class MockAssetWriter: NSObject, AssetWriter {
  var statusStub: (() -> AVAssetWriter.Status)?
  var startWritingStub: (() -> Bool)?
  var finishWritingStub: ((() -> Void) -> Void)?

  var status: AVAssetWriter.Status {
    return statusStub?() ?? .unknown
  }

  var error: Error? = nil

  func startWriting() -> Bool {
    return startWritingStub?() ?? true
  }

  func finishWriting(completionHandler handler: @escaping () -> Void) {
    finishWritingStub?(handler)
  }

  func startSession(atSourceTime startTime: CMTime) {}

  func add(_ input: AVAssetWriterInput) {}
}
